Observer/SalesQuoteProductAddAfter.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+
+namespace Klaviyo\Reclaim\Observer;
+
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\Catalog\Model\CategoryFactory;
+use Magento\Checkout\Model\Session;
+
+class SalesQuoteProductAddAfter implements ObserverInterface
+{
+    /**
+     * Magento Category Factory
+     * @var CategoryFactory $_categoryFactory
+     */
+    protected $_categoryFactory;
+
+    /**
+     * Magento Checkout Session data
+     * @var Session $_checkoutSession
+     */
+    protected $_checkoutSession;
+
+    /**
+     * SalesQuoteProductAddAfter constructor.
+     * @param CategoryFactory $categoryFactory
+     * @param Session $checkoutsession
+     */
+    public function __construct(
+        CategoryFactory $categoryFactory,
+        Session $checkoutsession
+    )
+    {
+        $this->_categoryFactory = $categoryFactory;
+        $this->_checkoutSession = $checkoutsession;
+    }
+
+    public function execute(Observer $observer)
+    {
+        $quote = $observer->getData('items')[0]->getQuote();
+        $addedItems = $observer->getData('items');
+
+        foreach ($addedItems as $item){
+            $this->klAddedToCartItemData($quote, $item);
+        }
+    }
+
+    /**
+     * Build Added to Cart payload for current Added item
+     * @param $quote
+     * @param $addedItem
+     */
+    public function klAddedToCartItemData($quote, $addedItem)
+    {
+        $addedProduct = $addedItem->getProduct();
+        $addedItemData = array(
+            'AddedItemCategories' => (array) $this->getCategoryName($addedProduct->getCategoryIds()),
+            'AddedItemDescription' => (string) strip_tags($addedProduct->getDescription()),
+            'AddedItemImageUrlKey' => (string) $addedProduct->getData('small_image'),
+            'AddedItemPrice' => (float) $addedProduct->get_price(),
+            'AddedItemQuantity' => (int) $addedItem->getQty(),
+            'AddedItemProductID' => (int) $addedProduct->getId(),
+            'AddedItemProductName' => (string) $addedItem->getName(),
+            'AddedItemSku' => (string) $addedProduct->getSku(),
+            'AddedItemUrl' => (string) $addedProduct->getProductUrl()
+        );
+
+        $klAddedToCartPayload = array_merge(
+            $this->klBuildCartData( $quote, $addedItem ),
+            $addedItemData
+        );
+
+        // Creating a custom session variable in the checkout session since quoteId is not set at this point for guest checkouts
+        $this->getCheckoutSession()->setKlAddedToCartKey( $klAddedToCartPayload );
```

Comment:
I think It's not the best option since after magento 2.4.3 and 2.3.7-p1 we have limit for the session size. Is there any other options?

Reply:
Hi @siddwarkhedkar,
The "add to cart" and "save quote after" are done in a single web request if I got it right.
In this case, it's better to introduce some object that will store this information and won't save data to session and DB till you need it.

Example of such code:
```php
class EventsQueue {
    private $items;

    public function addItem(array $payload): void
    {
        $this->items[] = $payload;
    }

    public function getItems(): array
    {
        return $this->items;
    }
}
```

And client code will be like this:

```php
// in the SalesQuoteProductAddAfter.php
$this->eventsQueue->addItem($klAddedToCartPayload);
```

and
```php
// in the SalesQuoteSaveAfter.php
$this->eventsQueue->getItems();
```